{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib import cm as cm\n",
    "from scipy.stats import binned_statistic\n",
    "from scipy.stats import sem\n",
    "import random\n",
    "import copy\n",
    "import warnings\n",
    "import plot_utils as plu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": false
   },
   "source": [
    "## LD \n",
    "Throughout the notebook, LD is measured by r^2 (the classical correlation coefficient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"* Computing and plotting LD...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute correlation between all pairs of SNPs for each generated/real dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# categ = infiles.keys()\n",
    "# hcor_snp = dict()\n",
    "# for i,cat in enumerate(categ):\n",
    "#     print(cat)\n",
    "#     with np.errstate(divide='ignore', invalid='ignore'): \n",
    "#         # Catch warnings due to fixed sites in dataset (the correlation value will be np.nan for pairs involving these sites)\n",
    "#         hcor_snp[cat] = np.corrcoef(datasets[cat], rowvar=False)**2  # r2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "### Plot LD or COV binned by distance between SNPs \n",
    "- LD is binned into 'nbins'\n",
    "- that are cut on a logscale if logscale=True\n",
    "- For regressions plot of non-binned LD, we randomly subsample 'nsamplesets' pairs for computation/visualization convenience\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# _, region_len, snps_on_same_chrom = plu.get_dist(position_fname['Real'], region_len_only=True,  kept_preprocessing=keptsnpdic['Real'])\n",
    "\n",
    "# nbins=50\n",
    "# nsamplesets=10000\n",
    "# logscale=True\n",
    "# bins = nbins\n",
    "# binsPerDist = nbins\n",
    "# if logscale: binsPerDist = np.logspace(np.log(1), np.log(region_len), nbins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Compute LD binned by distance\n",
    "# # Take only sites that are SNPs in all datasets (intersect)\n",
    "# # (eg intersection of SNPs in Real, SNPs in GAN, SNPs in RBM etc)\n",
    "# # -> Makes sense only if there is a correspondence between sites\n",
    "\n",
    "# if matching_SNPs:    \n",
    "#     binnedLD = dict()\n",
    "#     binnedPerDistLD = dict()\n",
    "#     kept_snp = ~is_fixed\n",
    "#     n_kept_snp = np.sum(kept_snp)\n",
    "#     realdist = plu.get_dist(position_fname['Real'], kept_preprocessing=keptsnpdic['Real'], kept_snp=kept_snp)[0]\n",
    "#     mat = hcor_snp['Real']\n",
    "#     # filter and flatten\n",
    "#     flatreal = (mat[np.ix_(kept_snp,kept_snp)])[np.triu_indices(n_kept_snp)]\n",
    "#     isnanReal = np.isnan(flatreal)\n",
    "#     i=1\n",
    "#     realdist = plu.get_dist(position_fname['Real'], kept_preprocessing=keptsnpdic['Real'], kept_snp=kept_snp)[0]\n",
    "#     hcor_snp_ref = hcor_snp.pop('Real')\n",
    "#     fig, axs = plt.subplots(nrows=2, ncols=len(hcor_snp), \n",
    "#                         figsize = (len(hcor_snp)*5, 2*5), constrained_layout=True)\n",
    "#     for ite, (cat, mat) in enumerate(hcor_snp.items()):\n",
    "#         print(ite)\n",
    "#         flathcor = (mat[np.ix_(kept_snp,kept_snp)])[np.triu_indices(n_kept_snp)]\n",
    "#         isnan = np.isnan(flathcor)\n",
    "#         curr_dist = plu.get_dist(position_fname[cat],  kept_preprocessing=keptsnpdic[cat], kept_snp=kept_snp)[0]\n",
    "\n",
    "#         # For each dataset LD pairs are stratified by SNP distance and cut into 'nbins' bins\n",
    "#         # bin per SNP distance\n",
    "#         ld = binned_statistic(curr_dist[~isnan], flathcor[~isnan], statistic = 'mean', bins=binsPerDist)\n",
    "#         with warnings.catch_warnings():\n",
    "#             warnings.simplefilter(\"ignore\", category=RuntimeWarning) # so that empty bins do not raise a warning\n",
    "#             binnedPerDistLD[cat] = pd.DataFrame({'bin_edges':ld.bin_edges[:-1],\n",
    "#                                           'LD': ld.statistic,\n",
    "#                                           #'sd': binned_statistic(curr_dist[~isnan], flathcor[~isnan], statistic = 'std', bins=binsPerDist).statistic,\n",
    "#                                           'sem': binned_statistic(curr_dist[~isnan], flathcor[~isnan], statistic = sem, bins=binsPerDist).statistic,\n",
    "#                                           'cat': cat, 'logscale': logscale})\n",
    "\n",
    "#         # For each dataset LD pairs are stratified by LD values in Real and cut into 'nbins' bins\n",
    "#         # binnedLD contains the average, std of LD values in each bin\n",
    "#         isnan = np.isnan(flathcor) |  np.isnan(flatreal) \n",
    "#         ld = binned_statistic(flatreal[~isnan], flathcor[~isnan], statistic = 'mean', bins=bins)\n",
    "#         with warnings.catch_warnings():\n",
    "#             warnings.simplefilter(\"ignore\", category=RuntimeWarning) # so that empty bins do not raise a warning\n",
    "#             binnedLD[cat] = pd.DataFrame({'bin_edges':ld.bin_edges[:-1],\n",
    "#                                           'LD': ld.statistic,\n",
    "#                                           'sd': binned_statistic(flatreal[~isnan], flathcor[~isnan], statistic = 'std', bins=bins).statistic,\n",
    "#                                           'sem': binned_statistic(flatreal[~isnan], flathcor[~isnan], statistic = sem, bins=bins).statistic,\n",
    "#                                           'cat': cat, 'logscale': logscale})\n",
    "\n",
    "#         # Plotting quantiles ?\n",
    "#         plu.plotregquant(x=flatreal, y=flathcor, \n",
    "#                          keys=['Truth',cat], statname='LD', col=colpal[cat], \n",
    "#                          step=0.0001,\n",
    "#                          ax=axs[1][ite])\n",
    "#         axs[1][ite].set_title(str(f'Quantiles LD {cat} vs Truth'))\n",
    "\n",
    "#         if matching_SNPs:\n",
    "#             # removing nan values and subsampling before doing the regression to have a reasonnable number of points\n",
    "#             isnanInter = isnanReal | isnan\n",
    "#             keepforplotreg = random.sample(list(np.where(~isnanInter)[0]), nsamplesets)\n",
    "#             plu.plotreg(x=flatreal[keepforplotreg], y=flathcor[keepforplotreg], \n",
    "#                         keys=['Truth',cat], statname='LD', col=colpal[cat], \n",
    "#                         ax=axs[0][ite])\n",
    "#             i+=1\n",
    "#             axs[0][ite].set_title(f'LD {cat} vs Truth')\n",
    "#     plt.savefig(outDir + \"LD_generated_vs_real_intersectSNP.pdf\", dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Plot LD as a fonction of binned distances\n",
    "# # except when SNPs are spread accross different chromosomes\n",
    "# if matching_SNPs: #(position_fname['Real']!=\"1kg_real/805snps.legend\"):\n",
    "#     tmp_real = colpal.pop('Real')\n",
    "#     print(colpal)\n",
    "#     sns.set_palette(colpal.values())\n",
    "\n",
    "#     sns.palplot(sns.color_palette())\n",
    "#     plt.figure(figsize=(7,7))\n",
    "#     # binnedPerDistLD.pop('Real')\n",
    "#     print(binnedPerDistLD.keys())\n",
    "#     for cat, bld in binnedPerDistLD.items():\n",
    "#         plt.errorbar(bld.bin_edges.values, bld.LD.values, bld['sem'].values, label=cat, alpha=.65,linewidth=3)\n",
    "#     plt.title(\"Binned LD +/- 1 sem\")\n",
    "#     if (logscale): plt.xscale('log')   \n",
    "#     #plt.yscale('log')\n",
    "#     plt.xlabel(\"Distance between SNPs (bp) [Left bound of distance bin]\")\n",
    "#     plt.ylabel(\"Average LD in bin\")\n",
    "#     plt.legend()\n",
    "#     plt.savefig(outDir + \"correlation_vs_dist_intersectSNP.pdf\", dpi=300, bbox_inches='tight')\n",
    "#     colpal['Real'] = tmp_real\n",
    "\n",
    "#     # Zoomed-in plot for short distances (1 to 1000 bp)\n",
    "#     plt.figure(figsize=(7,7))\n",
    "#     for cat, bld in binnedPerDistLD.items():\n",
    "#         plt.errorbar(bld.bin_edges.values, bld.LD.values, bld['sem'].values, label=cat, alpha=.65, linewidth=3)\n",
    "#     plt.title(\"Binned LD +/- 1 sem (Zoom: 1–1000 bp)\")\n",
    "#     plt.xscale('log')\n",
    "#     plt.xlim(1, 1000)  # Focus on 10^0 to 10^3\n",
    "#     plt.xlabel(\"Distance between SNPs (bp) [Left bound of distance bin]\")\n",
    "#     plt.ylabel(\"Average LD in bin\")\n",
    "#     plt.legend()\n",
    "#     plt.savefig(outDir + \"correlation_vs_dist_zoomed_1_1000bp.pdf\", dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import binned_statistic, sem\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning)\n",
    "\n",
    "def load_positions(legend_file):\n",
    "    \"\"\"Load SNP genomic positions from legend file (1D array of ints).\"\"\"\n",
    "    df = pd.read_csv(legend_file, sep=\" \")\n",
    "    return df[\"position\"].values.astype(int)\n",
    "\n",
    "def compute_ld_r2(geno):\n",
    "    \"\"\"Compute r² LD matrix from genotype matrix (n_samples × n_snps).\"\"\"\n",
    "    with np.errstate(divide=\"ignore\", invalid=\"ignore\"):\n",
    "        return np.corrcoef(geno, rowvar=False)**2\n",
    "\n",
    "def flatten_upper(mat):\n",
    "    \"\"\"Return flattened upper triangle of a square matrix (no diagonal).\"\"\"\n",
    "    iu = np.triu_indices(mat.shape[0], k=1)\n",
    "    return mat[iu]\n",
    "\n",
    "def compute_pairwise_distances(positions):\n",
    "    \"\"\"Compute pairwise SNP distances as flattened upper triangle.\"\"\"\n",
    "    diff = np.abs(positions[:, None] - positions[None, :])\n",
    "    iu = np.triu_indices(len(positions), k=1)\n",
    "    return diff[iu]\n",
    "\n",
    "def compute_binned_ld(dists, ld, nbins=50, logscale=True):\n",
    "    \"\"\"Bin LD by distance and compute mean + SEM.\"\"\"\n",
    "    valid = ~np.isnan(ld)\n",
    "    d = dists[valid]\n",
    "    l = ld[valid]\n",
    "\n",
    "    if logscale:\n",
    "        bins = np.logspace(np.log10(max(1, d.min())), np.log10(d.max()), nbins)\n",
    "    else:\n",
    "        bins = nbins\n",
    "\n",
    "    ld_mean = binned_statistic(d, l, statistic=\"mean\", bins=bins)\n",
    "    ld_sem  = binned_statistic(d, l, statistic=sem, bins=bins)\n",
    "\n",
    "    return pd.DataFrame({\n",
    "        \"bin_left\": ld_mean.bin_edges[:-1],\n",
    "        \"LD_mean\": ld_mean.statistic,\n",
    "        \"LD_sem\": ld_sem.statistic\n",
    "    })\n",
    "\n",
    "# --- small helper: safe color lookup ---\n",
    "def get_color(name, colpal, default=\"#333333\"):\n",
    "    \"\"\"Return color for name from colpal, or default if not present.\"\"\"\n",
    "    return colpal.get(name, default)\n",
    "\n",
    "def plot_ld_decay_with_colors(binned, colpal, outfile, logscale=True, zoom=None):\n",
    "    plt.figure(figsize=(7, 7))\n",
    "    for label in plot_order:\n",
    "        if label not in binned:\n",
    "            continue\n",
    "        df = binned[label]\n",
    "        color = get_color(label, colpal)\n",
    "        plt.errorbar(df.bin_left, df.LD_mean, df.LD_sem,\n",
    "                     label=label, alpha=0.8, linewidth=3, color=color)\n",
    "\n",
    "    if logscale:\n",
    "        plt.xscale(\"log\")\n",
    "    if zoom:\n",
    "        plt.xlim(*zoom)\n",
    "\n",
    "    plt.xlabel(\"Distance between SNPs (bp)\")\n",
    "    plt.ylabel(\"Average LD (r²)\")\n",
    "    # plt.title(\"LD decay (+/- 1 SEM)\")\n",
    "\n",
    "    # Legend with robust linewidth handling\n",
    "    leg = plt.legend()\n",
    "    handles = getattr(leg, \"legendHandles\", None)\n",
    "    if handles is None:\n",
    "        handles = getattr(leg, \"legend_handles\", [])\n",
    "    for h in handles:\n",
    "        try:\n",
    "            h.set_linewidth(3.0)\n",
    "        except Exception:\n",
    "            pass\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(outfile, dpi=300)\n",
    "    plt.close()\n",
    "\n",
    "# Step 0: load positions and filter out fixed SNPs\n",
    "positions_all = load_positions(realposfname)\n",
    "kept_snp_mask = ~is_fixed\n",
    "positions = positions_all[kept_snp_mask]\n",
    "\n",
    "print(f\"Number of unfixed SNPs: {len(positions)}\")\n",
    "\n",
    "# Step 1: compute LD (filtered)\n",
    "ld_matrices = {\n",
    "    name: compute_ld_r2(g[:, kept_snp_mask])\n",
    "    for name, g in datasets.items()\n",
    "}\n",
    "\n",
    "# Step 2: flatten LD\n",
    "flat_ld = {name: flatten_upper(ld_matrices[name]) for name in ld_matrices}\n",
    "\n",
    "# Step 3: compute pairwise SNP distances with same filter\n",
    "pairwise_distances = compute_pairwise_distances(positions)\n",
    "\n",
    "# Step 4: compute LD decay curves\n",
    "binned_ld = {\n",
    "    name: compute_binned_ld(pairwise_distances, flat_ld[name], nbins=20)\n",
    "    for name in ld_matrices\n",
    "}\n",
    "\n",
    "# Exclude Real from plotted lines (but keep it in binned_ld for reference)\n",
    "binned_ld_no_real = {k: v for k, v in binned_ld.items() if k != \"Real\"}\n",
    "\n",
    "# Create a deterministic plot order (optional: preserve the original order in infiles.keys())\n",
    "plot_order = [k for k in infiles.keys() if (k in binned_ld_no_real)]\n",
    "# if some keys in binned_ld_no_real are not in infiles, append them\n",
    "for k in binned_ld_no_real:\n",
    "    if k not in plot_order:\n",
    "        plot_order.append(k)\n",
    "\n",
    "# Now call the plotting function (Real excluded)\n",
    "plot_ld_decay_with_colors(\n",
    "    binned_ld_no_real,\n",
    "    colpal,\n",
    "    outfile=outDir + \"LD_decay.pdf\",\n",
    "    logscale=True\n",
    ")\n",
    "\n",
    "plot_ld_decay_with_colors(\n",
    "    binned_ld_no_real,\n",
    "    colpal,\n",
    "    outfile=outDir + \"LD_decay_zoom_1_1000bp.pdf\",\n",
    "    logscale=True,\n",
    "    zoom=(1, 1000)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each dataset LD pairs were stratified by LD values in Real, cut into nbins bins\n",
    "# binnedLD contains the average LD in each bin\n",
    "# Plot generated average LD as a function of the real average LD in the bins\n",
    "# if matching_SNPs:\n",
    "#     plt.figure(figsize=(10,10))\n",
    "#     for cat, bld in binnedLD.items():\n",
    "#         plt.errorbar(bld.bin_edges.values, bld.LD.values, bld['sem'].values, label=cat, alpha=1, marker='o')\n",
    "#     plt.title(\"Binned LD +/- 1 sem\")\n",
    "#     #if (logscale): plt.xscale('log')\n",
    "#     plt.xlabel(\"Bins (LD in Real)\")\n",
    "#     plt.ylabel(\"Average LD in bin\")\n",
    "#     plt.legend()\n",
    "#     plt.savefig(outDir+'LD_{}bins_{}fixedremoved.pdf'.format(nbins,'logdist_' if logscale else ''))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plotting LD (block) matrix\n",
    "- LD is r2\n",
    "- if mirror=True plot the symmetrical matrix for each category\n",
    "- if mirror=False plot the Generated LD in upper triangle versus the Real LD in lower triangle\n",
    "- if diff=True plot the difference between generated and true LD values, else plot regular LD values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(\"* Plotting LD block matrices...\")\n",
    "\n",
    "# # Set edges of the region for which to plot LD block matrix (l=0, f='end') for full region\n",
    "# # not used as for now apart from the filename\n",
    "# l_bound=None\n",
    "# r_bound=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # mirror (bool): plot symmetrical matrix or generated vs real?\n",
    "# # diff (bool): plot LD values or generated minus real ?\n",
    "\n",
    "# # (mirror=False, diff=False) = upper triangle = generated LD, lower triangle = real LD, raw values from 0 to 1.\n",
    "\n",
    "# hcor_snp['Real'] = hcor_snp_ref\n",
    "# print(hcor_snp.keys())\n",
    "# # for snpcode in (\"fullSNP\", \"intersectSNP\"):\n",
    "# #     for (mirror, diff) in ((True,False),(True,True), (False,False)):\n",
    "# hcor_snp.pop('WGAN')\n",
    "# hcor_snp.pop('RBM')\n",
    "# hcor_snp.pop('Indep')\n",
    "# hcor_snp.pop('Markov')\n",
    "# hcor_snp.pop('HMM')\n",
    "# hcor_snp.pop('Truth')\n",
    "\n",
    "# for snpcode in (\"fullSNP\", \"intersectSNP\"):\n",
    "#     for (mirror, diff) in ((True,False),(True,True),(False,False)):\n",
    "#         if (not matching_SNPs) and (snpcode==\"fullSNP\") and (diff or not mirror):\n",
    "#             print(f'Warning: not plotting LD for {snpcode} mirror={mirror} diff={diff}',\n",
    "#                   ' because SNP have no one-to-one correspondence and matrices might have different sizes')\n",
    "#         elif (matching_SNPs) and (snpcode==\"fullSNP\"):\n",
    "#             pass\n",
    "#         else:\n",
    "#             outfilename = f\"LD_HEATMAP_{snpcode}_bounds={l_bound}-{r_bound}_mirror={mirror}_diff={diff}.pdf\" \n",
    "#             #print(outfilename)\n",
    "#             fig = plt.figure(figsize=(7,7))\n",
    "#             plu.plotLDblock(hcor_snp, \n",
    "#                             left=l_bound, right=r_bound,  # None, None -> takes all SNPs\n",
    "#                             mirror=mirror, diff=diff,\n",
    "#                             is_fixed=is_fixed, is_fixed_dic=is_fixed_dic,\n",
    "#                             suptitle_kws={'t':outfilename}\n",
    "#                            )\n",
    "#             plt.tight_layout()\n",
    "#             plt.savefig(outDir+outfilename, dpi=300, bbox_inches='tight')\n",
    "#             plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('****************************************************************\\n*** Computation and plotting LD DONE. Figures saved in {} ***\\n****************************************************************'.format(outDir))"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "fd79f911c2d45fc4418704ce15f7296bc75811b04b5135772199ee1ccc3f6b00"
  },
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc-autonumbering": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
